File: app/src/main/java/example/develop/davidoh/java_android_mvp_example/data/source/image/ImageRepository.java
package example.develop.davidoh.java_android_mvp_example.data.source.image;

public class ImageRepository implements ImageDatasource {
    private static ImageRepository imageRepository;
    private ImageLocalData imageLocalData;
    private ImageRemoteData imageRemoteData;

    public static ImageRepository getInstance() {
        if (imageRepository == null){
            synchronized (ImageRepository.class) {
                if (imageRepository == null) {
                    imageRepository = new ImageRepository();
                }
            }
        }
        return imageRepository;
    }

    private ImageRepository() {
        imageLocalData = ImageLocalData.getInstance();
        imageRemoteData = ImageRemoteData.getInstance();
    }


    @Override
    public void loadImageList(LoadImageCallback loadImageCallback, int size) {
        //imageLocalData.loadImageList(loadImageCallback, size);
        imageRemoteData.loadImageList(loadImageCallback, size);
    }
}

File: app/src/main/java/example/develop/davidoh/java_android_mvp_example/data/source/image/ImageRepository.kt
package example.develop.davidoh.java_android_mvp_example.data.source.image

import example.develop.davidoh.java_android_mvp_example.data.ImageData

class ImageRepository : ImageDataSource {
    private val IMAGE_REMOTE_DATA: ImageRemoteData by lazy {
        ImageRemoteData()
    }

    override fun loadImageList(imageDataList: (List<ImageData>) -> Unit, size: Int) {
        IMAGE_REMOTE_DATA.loadImageList(imageDataList, size)
    }
}
